File: BraintreeModels/src/test/java/com/braintreepayments/api/models/CardConfigurationUnitTest.java
package com.braintreepayments.api.models;

import com.braintreepayments.testutils.Fixtures;

import org.json.JSONException;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class CardConfigurationUnitTest {

    @Test
    public void parsesCardConfiguration() throws JSONException {
        Configuration configuration = Configuration.fromJson(Fixtures.CONFIGURATION_WITH_SUPPORTED_CARD_TYPES);
        CardConfiguration cardConfiguration = configuration.getCardConfiguration();

        assertEquals(5, cardConfiguration.getSupportedCardTypes().size());
        assertTrue(cardConfiguration.getSupportedCardTypes().contains("American Express"));
        assertTrue(cardConfiguration.getSupportedCardTypes().contains("Discover"));
        assertTrue(cardConfiguration.getSupportedCardTypes().contains("JCB"));
        assertTrue(cardConfiguration.getSupportedCardTypes().contains("MasterCard"));
        assertTrue(cardConfiguration.getSupportedCardTypes().contains("Visa"));
        assertFalse(cardConfiguration.isFraudDataCollectionEnabled());
    }

    @Test
    public void handlesNull() {
        CardConfiguration cardConfiguration = CardConfiguration.fromJson(null);

        assertEquals(0, cardConfiguration.getSupportedCardTypes().size());
    }

    @Test
    public void isFraudDataCollectionEnabled_isTrueWhenEnabled() throws JSONException {
        Configuration configuration = Configuration.fromJson(Fixtures.CONFIGURATION_WITH_CARD_COLLECT_DEVICE_DATA);
        CardConfiguration cardConfiguration = configuration.getCardConfiguration();

        assertTrue(cardConfiguration.isFraudDataCollectionEnabled());
    }
}

File: BraintreeCore/src/test/java/com/braintreepayments/api/core/CardConfigurationUnitTest.kt
package com.braintreepayments.api.core

import org.robolectric.RobolectricTestRunner
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(RobolectricTestRunner::class)
class CardConfigurationUnitTest {
    @Test
    fun `when input contains full card configuration data, all fields are parsed`() {
        val input = JSONObject()
            .put("collectDeviceData", true)
            .put(
                "supportedCardTypes", JSONArray()
                    .put("American Express")
                    .put("Discover")
                    .put("JCB")
                    .put("MasterCard")
                    .put("Visa")
            )
        val (supportedCardTypes, isFraudDataCollectionEnabled) = CardConfiguration(input)
        assertTrue(isFraudDataCollectionEnabled)
        assertEquals(5, supportedCardTypes.size)
        assertEquals("American Express", supportedCardTypes[0])
        assertEquals("Discover", supportedCardTypes[1])
        assertEquals("JCB", supportedCardTypes[2])
        assertEquals("MasterCard", supportedCardTypes[3])
        assertEquals("Visa", supportedCardTypes[4])
    }

    @Test
    fun `when input is null, CardConfiguration uses default values`() {
        val (supportedCardTypes, isFraudDataCollectionEnabled) = CardConfiguration(null)
        assertFalse(isFraudDataCollectionEnabled)
        assertEquals(0, supportedCardTypes.size)
    }

    @Test
    fun `when input is empty, CardConfiguration uses default values`() {
        val (supportedCardTypes, isFraudDataCollectionEnabled) = CardConfiguration(JSONObject())
        assertFalse(isFraudDataCollectionEnabled)
        assertEquals(0, supportedCardTypes.size)
    }
}
